rfd/0201-native-auto-approval.md

```diff
@@ -0,0 +1,318 @@
+    contains_any(user.traits["location"], set("Seattle"))
+  automatic_approval:
+    name: teleport
+  notification:
+    name: slack
+    recipients: ["#dev-cloud"]
+```
+
+The `Access Monitoring Rules` overview page will be modified to display both
+notification rules, as well as automatic approval rules. This page will allow
+user to see a quick overview of the automatic approvals currently enabled. The
+overview simply displays the access monitoring rule name, plugin/integration
+name, and the roles that are automatically approved. Users will need to click
+on the **View** button to see the actual conditions for auto approval.
+
+![view-amr](assets/0201-view-amr.png)
+
+## Details
+This feature will be supported by a new internal Teleport service. This
+automatic approval service will function similarly to existing access plugins.
+It will be running as part of the Teleport Auth Service by default.
+
+The automatic approval service relies on a similar workflow that supports access
+request notification routing with access monitoring rules. The service watches
+for Access Monitoring (AMR) events and Access Request (AR) events. If an
+incoming AR matches an existing AMR condition, then the service will attempt to
+automatically approve the request.
+
+### Access Monitoring Rule
+The AMR now supports a configurable `automatic_approval.name` spec. This field
+specifies the plugin/service responsible for handling automatic approvals for
+matching ARs. For the initial implementation, the only acceptable value will be
+"teleport", indicating that the internal Teleport service will handle the
+request. This field will eventually accept values for all the available access
+plugins that support automatic approvals.
+
+The AMR conditions now also support a `user.traits` variable.  This variable
+maps a trait name to a set of values. This allows users to specify arbitrary
+user traits, such as, "level", "team", "location", etc...  These traits can then
+be used to identity whether a user is on-call, or if the user is pre-approved
+for the access request.
+
+The AMR conditions now also support the `contains_all(list, items)` operator.
+This operator returns true if `list` contains an exact match for all elements of
+`items`. This operator allows users to configure a more restrictive condition
+for automatic approvals.
+
+Additionally, the `spec.states` field will now be utilized to support an
+`approved` state. If the AMR contains `spec.states.approved`, this indicates
+that the AR should be automatically approved.
+
+```yaml
+# This AMR would allow users with traits "level"="L1" and "team"="Cloud" and
+# "location"="Seattle" to be pre-approved for the "cloud-dev" role.
+kind: access_monitoring_rule
+version: v1
+metadata:
+  name: cloud-dev-pre-approved
+spec:
+  subjects:
+    - access_request
+  # states.approved indicates that matching access requests should be automatically approved.
+  states:
+    - approved
```

Comment:
We could support automatic rejections as well but not sure there's a use-case for it yet. Might wanna mention it as a potential future work at least.